src/main/java/org/junit/internal/runners/statements/FailOnTimeout.java

```diff
@@ -122,12 +122,22 @@ public void evaluate() throws Throwable {
         FutureTask<Throwable> task = new FutureTask<Throwable>(callable);
         ThreadGroup threadGroup = new ThreadGroup("FailOnTimeoutGroup");
         Thread thread = new Thread(threadGroup, task, "Time-limited test");
-        thread.setDaemon(true);
-        thread.start();
-        callable.awaitStarted();
-        Throwable throwable = getResult(task, thread);
-        if (throwable != null) {
-            throw throwable;
+        try {
+            thread.setDaemon(true);
+            thread.start();
+            callable.awaitStarted();
+            Throwable throwable = getResult(task, thread);
+            if (throwable != null) {
+                throw throwable;
+            }
+        } finally {
+            thread.join(100);
```

Comment:
Any reason why we can't pass 1 to `thread.join()`? At this point, the test either completed (in which case this should return right away) or the test timed out (in which case the thread will likely take more than 100ms to exit). Waiting extra time just makes the test run longer.

In fact, maybe we shouldn't call join at all. thread.join() can throw `InterruptedException`.

Reply:
I reduced the timeout as suggested.

`join` has to be called because otherwise the thread will still be active when calling `threadGroup.destroy();`, which will throw an `IllegalThreadStateException` and keep the group referenced, causing the leak.